### PR TITLE
UCT/UGNI: move misplaced functions out of ugni_iface.c

### PR DESCRIPTION
**Pull request by MattBBaker**
Very little new code here, mostly moving existing code. The objective was to make make the job_info struct private to ugni_device.c and no where else in the code depends on it. While doing surgery on some files it also moves some rdma specific code out of the base iface and puts it where it belongs.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1713/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3650/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3652/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1723/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3663/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@yosefe  
hmm what are these build failure

File: src/uct/ugni/base/ugni_types.h [hunk 1]
```diff
@@ -17,13 +17,10 @@ typedef struct uct_ugni_device {
     char             type_name[UCT_UGNI_MAX_TYPE_NAME];  /**< Device type name */
     char             fname[UCT_DEVICE_NAME_MAX];/**< Device full name */
     uint32_t         device_id;                 /**< Device id */
-    int              device_index;              /**< Index of the device in the
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_types.h]: Any particular reason why you have removed it ?

File: src/uct/ugni/base/ugni_types.h [hunk 2]
```diff
@@ -17,13 +17,10 @@ typedef struct uct_ugni_device {
...
-                                                  array of devices */
     uint32_t         address;                   /**< Device address */
     uint32_t         cpu_id;                    /**< CPU attached directly
                                                   to the device */
     cpu_set_t        cpu_mask;                  /**< CPU mask */
-    bool             attached;                  /**< device was attached */
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_types.h]: The same here

File: src/uct/ugni/base/ugni_types.h (hunk 1 above)
Comment (MattBBaker) [inline src/uct/ugni/base/ugni_types.h]: After the refactoring it was unused. Removed to reclaim a few bytes from the struct.

File: src/uct/ugni/base/ugni_types.h (hunk 2 above)
Comment (MattBBaker) [inline src/uct/ugni/base/ugni_types.h]: I don't think this was ever used. So removed because it was set once to false and forgotten.

**Comment by shamisp** [comment]
On Cray machines by default there is only one NIC. When I designed it I thought It would be useful to have conceptual support for multi-nic... just in case. I would say for now it is up to , if you want to preserve multi-nic support.

**Comment by MattBBaker** [comment]
At this point I'm not going to put special effort to to support multi nic with cray asics. I don't think the ugni api actually supports it either. Not a priority to remove it.

**Comment by shamisp** [comment]
Okay - I don't have strong opinion, but just wanted to highlight this design point.

**Comment by shamisp** [comment]
Seems like Mellanox lab failure is irrelevant.

**Pull request merged**